File: moderate/chain_inspection.rb
File.open(ARGV[0]).each_line do |line|
  b = 'BEGIN'
  s = line.chomp.split(';')
  m = {}
  been = {}
  s.each do |x|
    t = x.split('-')
    m[t[0]] = t[1]
  end
  s.length.times do
    if !m.key?(b) || been.key?(m[b])
      b = nil
      break
    end
    b = m[b]
    been[b] = true
  end
  puts b == 'END' ? 'GOOD' : 'BAD'
end

File: moderate/chain_inspection.lua
for line in io.lines(arg[1]) do
  n, b, m, been = 0, "BEGIN", {}, {}
  for i in line:gmatch("[^;]+") do
    t, u = i:match("([^-]+)-([^-]+)")
    m[t], n = u, n + 1
  end
  for i = 1, n do
    b = m[b]
    if not b or been[b] then
      b = nil
      break
    end
    been[b] = true
  end
  print((b == "END") and "GOOD" or "BAD")
end
